pennylane/devices/qubit/apply_operation.py

```diff
@@ -228,9 +229,27 @@ def _(op: type_op, state):
         [1., 0.]], requires_grad=True)
 
     """
+    if sp.sparse.issparse(op.matrix()):
+        return apply_operation_csr_matrix(op, state, is_state_batched)
     return _apply_operation_default(op, state, is_state_batched, debugger)
 
 
+def apply_operation_csr_matrix(op, state, is_state_batched: bool = False):
+    """The csr_matrix specialized version apply operation."""
+    # Calculate the num wires by state shape
+    if sp.sparse.issparse(state):  # Then the first is batch and the second is state dim
+        len_state = state.shape[1]
+        num_wires = int(np.log2(len_state))
+        return state @ op.matrix(wire_order=range(num_wires)).T
```

Comment:
At this point, we would have a sparse vector, but are we asking for the operator dense representation?

Reply:
@AmintorDusko The `op` here should be a sparse and its `op.matrix()` should also return a sparse as we did in `QubitUntiary` (though right now this is the only one that has such feature)

Reply:
Hmm. Make sense, but if in cases where the OP is something defined by us?
PauliX for example.
Shouldn't we have a name for dense and a name for sparse representations?

Reply:
So like `PauliX_sparse`?

Reply:
yes

Reply:
@AmintorDusko sorry somehow I totally forgot this thread... Actually right now even though the default `matrix()` method of `Pauli` still gives out the dense matrix nparray, its `sparse_matrix()` method will work well and return a CSR. Is this good enough or we should have a `PauliX_sparse` whose `matrix()` just return the CSR as before?

Reply:
It is weird to have a function returning dense or sparse in different situations. Perhaps we should only output a sparse matrix when we call `sparse_matrix()`, keeping `matrix()` for dense representation only. I'm happy to be convinced otherwise.

Reply:
Oh I see. Indeed this is a weird behavior now. How about we re-define the `QubitUnitary` accepting sparse matrices as input, but still convert to dense if `matrix()` is called? Then I can add another round of  specific conditions inside `apply_operation` to help tunnel whether or not to use sparse repr (e.g. additional kw str, or dynamically decide according to system size and sparsity).

Also, I see this previous added feature caused confusion for @albi3ro yesterday as well, so right now it might be the perfect time to discuss and make a decision.

Reply:
Continued converstaions can be found https://github.com/PennyLaneAI/pennylane/pull/6986